File: ToolKit/src/main/java/com/uday/android/toolkit/fragments/PartitionSchemeFragment.java
package com.uday.android.toolkit.fragments;

import android.annotation.*;
import android.app.*;
import android.content.*;
import android.graphics.*;
import android.graphics.drawable.*;
import android.os.*;
import android.util.*;
import android.view.*;
import android.view.View.*;
import android.widget.*;
import android.widget.LinearLayout.*;
import com.uday.android.toolkit.*;
import com.uday.android.util.*;
import eu.chainfire.libsuperuser.*;
import java.io.*;
import java.util.*;

import android.view.View.OnClickListener;
import android.support.v4.widget.*;
import com.uday.android.toolkit.ui.*;

@SuppressLint("NewApi")
public class PartitionSchemeFragment extends Fragment
{
	private Context context;
	private File DISK;
	private ListView partitionListView;
	private long DISK_SIZE;
	private PartitionListAdapter adapter;
	private RelativeLayout rootView;
	private String SECTOR_SIZE;
	private String STORAGE_MODEL;
	private String PARTITION_TABLE;
	private SwipeRefreshLayout partitionSwipeRefresh;
	private PartedParserRunnable partedParser;
	private View headerView;
	private List<String> mainBlockDevices;
	private ArrayAdapter spinnerAdapter;
	private final String blockDev0="/dev/block/mmcblk0";
	private final String blockDev1="/dev/block/mmcblk1";
	
	private static ArrayList<BlockDeviceListData> blockDevicesList;
	
	public PartitionSchemeFragment(Context context){
		this.context=context;
		partedParser=new PartedParserRunnable();
	}

	@Override
	public Context getContext()
	{
		if(Build.VERSION.SDK_INT<=Build.VERSION_CODES.M)
			return context;
		return super.getContext();
	}
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {
		if(rootView==null){
			rootView = (RelativeLayout)inflater
				.inflate(R.layout.partition_scheme, container, false);
			onViewFirstCreated();
		}
		
		rootView.startAnimation(((MainActivity)context).mGrowIn);
		return rootView;
	}
	
	private void onViewFirstCreated(){
		partitionSwipeRefresh=(SwipeRefreshLayout)rootView.findViewById(R.id.partition_swipe_refresh);
		partitionSwipeRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener(){
			@Override public void onRefresh(){
				partitionSwipeRefresh.setRefreshing(true);
				refresh(1);
			}
		});
		partitionListView=(ListView)rootView.findViewById(R.id.partition_list_view);
		if(blockDevicesList==null){
			partitionSwipeRefresh.setRefreshing(true);
			blockDevicesList=new ArrayList<BlockDeviceListData>();
			newAdapter();
			refresh(0);
		}else newAdapter();
		
		mainBlockDevices=new ArrayList<String>();
		if(new File(blockDev0).exists())mainBlockDevices.add(blockDev0);
		if(new File(blockDev1).exists())mainBlockDevices.add(blockDev1);
		partitionListView.setAdapter(adapter);
		headerView=((Activity)getContext()).getLayoutInflater().inflate(R.layout.partition_view_header,null);
		spinnerAdapter=new ArrayAdapter(getContext(),android.R.layout.simple_list_item_1,mainBlockDevices);
		((Spinner)headerView.findViewById(R.id.part_disk_spinner)).setAdapter(spinnerAdapter);
	
		headerView.setVisibility(View.GONE);
		partitionListView.addHeaderView(headerView);
	}
	
	private void newAdapter(){
		adapter=new PartitionListAdapter(getContext(),R.layout.partition_list_item,blockDevicesList);
	}
	
	private void refresh(final int type){
		blockDevicesList.clear();
		adapter.notifyDataSetChanged();
		if(!((MainActivity)getContext()).backgroundThreadisRunning){
			runInBackground(partedParser.setType(type));
		}else{
			new Thread(partedParser).start();
		}
	}

	private void onLoadingCompleted(String partedData){
		blockDevicesList.clear();
		blockDevicesList.addAll(parseParted(partedData));
		runOnUiThread(new Runnable(){
				@Override
				public void run(){
					headerView.setVisibility(View.VISIBLE);
					adapter.notifyDataSetChanged();
					partitionSwipeRefresh.setRefreshing(false);
					setHasOptionsMenu(true);
					mainBlockDevices.remove(0);
					mainBlockDevices.add(0,blockDev0+" : "+Utils.getConventionalSize(DISK_SIZE));
					spinnerAdapter.notifyDataSetChanged();
					((TextView)headerView.findViewById(R.id.part_table)).setText(PARTITION_TABLE);
					((TextView)headerView.findViewById(R.id.disk_model)).setText(STORAGE_MODEL);
					((TextView)headerView.findViewById(R.id.sector_size)).setText(SECTOR_SIZE);
				}
			});
	}
	
	private void runInBackground(Runnable action){
		((MainActivity)getContext()).runInBackground(action);
	}
	
	private void runOnUiThread(Runnable action){
		((MainActivity)getContext()).runOnUiThread(action);
	}
	
	private void writeToFile(String data,String fileName,Context context){
		FileOutputStream outputStream;
		try{
			outputStream =context.openFileOutput(fileName, Context.MODE_PRIVATE);
			outputStream.write(data.getBytes());
			outputStream.close();
		}catch(Exception ex){
			Log.e(MainActivity.TAG,"File write failed: "+ex.toString());
		} 
	}
	
	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		inflater.inflate(R.menu.menu_part_unit,menu);
		super.onCreateOptionsMenu(menu, inflater);
	}
	
	@Override
    public boolean onOptionsItemSelected(MenuItem item){
		for(BlockDeviceListData data:blockDevicesList)
			data.setSizeUnit(item.getItemId());
		adapter.notifyDataSetChanged();
		return super.onOptionsItemSelected(item);
	}
	
	private ArrayList<BlockDeviceListData> parseParted(String str){
		ArrayList<BlockDeviceListData> devicesList=new ArrayList<BlockDeviceListData>();
		for(String tmp: str.split("\n"))
		{
			if(STORAGE_MODEL==null && tmp.contains("Model"))
			{
				STORAGE_MODEL=tmp.replace("Model: ","").replace("Model","");
			}
			else if(tmp.contains("Disk "))
			{
				try{
					DISK=new File(tmp.split(":")[0].split(" ")[1]);
					DISK_SIZE=Long.parseLong(tmp.split(":")[1].replace("B","").replace(" ",""));
				}catch(NullPointerException ex){
					Log.e(MainActivity.TAG,ex.toString());
				}
			}
			else if(tmp.contains("Sector size"))
			{
				try{
					SECTOR_SIZE=tmp.split(":")[1].replace(" ","");
				}catch(NullPointerException ex){
					Log.e(MainActivity.TAG,ex.toString());
				}
			}
			else if(tmp.contains("Partition Table"))
			{
				try{
					PARTITION_TABLE=tmp.split(":")[1].replace(" ","");
				}catch(NullPointerException ex){
					Log.e(MainActivity.TAG,ex.toString());
				}
			}
			else if(tmp.contains("Number "))
				for(String test : str.split(tmp)[1].split("\n"))
					if(test.length()>10){
						int i=0;
						BlockDeviceListData data=new BlockDeviceListData();
						for(String another :test.split(" "))
							if(!another.equals("")){
								switch(i)
								{
									case 0:data.setBlock(new File(DISK+"p"+Integer.parseInt(another)));break;
									case 1:data.setStart(Long.parseLong(another.split("B")[0].toString()));break;
									case 2:data.setEnd(Long.parseLong(another.split("B")[0].toString()));break;
									case 3:data.setSize(Long.parseLong(another.split("B")[0].toString()));break;
									default:try{
												if(!(data.getName()==null))
													data.setType(data.getName());
												else data.setType("");
											}catch(Exception ex){
												data.setType("");
											}
											data.setName(another);break;
								}
								i++;
							}
						if(data.getBlock()!=null)
							devicesList.add(data);
					}
		}
		return devicesList;
	}
	
	private class PartedParserRunnable implements Runnable{
		private int type=0;

		public Runnable setType(int type){
			this.type=type;
			return this;
		}

		@Override
		public void run(){
			File dataFile=new File(getContext().getFilesDir()+"/partition_scheme.info");
			String scheme="";
			if(dataFile.exists() && type==0){
				try{
					String tmp="";
					BufferedReader reader=new BufferedReader(new InputStreamReader(new FileInputStream(dataFile)));
					while((tmp=reader.readLine())!=null){
						scheme+=tmp+"\n";
					}
					onLoadingCompleted(scheme);
				}catch(IOException ex){
					Log.e("Exception","Failed to read "+dataFile.getAbsolutePath(),ex);
				}
			}
			else{
				if(dataFile.exists())dataFile.delete();
				MainActivity.rootSession.addCommand(getContext().getFilesDir()+"/common/partition_scheme.sh "+MainActivity.TOOL+" 'b'",2323,new Shell.OnCommandResultListener(){
						@Override
						public void onCommandResult(int commandcode,int exitcode,List<String> output){
							writeToFile(Utils.getString(output),"partition_scheme.info",context);
							onLoadingCompleted(Utils.getString(output));
						}
					});		
			}
		}
	}
}

File: app/src/main/java/com/uday/android/toolkit/fragments/PartitionSchemeFragment.kt
package com.uday.android.toolkit.fragments

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.uday.android.toolkit.MainActivity
import com.uday.android.toolkit.R
import com.uday.android.toolkit.ui.PartitionListAdapter
import com.uday.android.util.BlockDeviceListData
import com.uday.android.util.Utils
import java.io.*
import java.util.*

@SuppressLint("NewApi", "ValidFragment")
class PartitionSchemeFragment @SuppressLint("ValidFragment")
    constructor(private val context:Context): androidx.fragment.app.Fragment() {
    private var disk:File? = null
    private var partitionListView:ListView? = null
    private var diskSize:Long = 0
    private var adapter:PartitionListAdapter? = null
    private var rootView:RelativeLayout? = null
    private var selectSize:String? = null
    private var storageModel:String? = null
    private var partitionTable:String? = null
    private var partitionSwipeRefresh: SwipeRefreshLayout? = null
    private val partedParser:PartedParserRunnable
    private var headerView:View? = null
    private var mainBlockDevices:MutableList<String>? = null
    private var spinnerAdapter:ArrayAdapter<*>? = null
    private val blockDev0 = "/dev/block/mmcblk0"
    private val blockDev1 = "/dev/block/mmcblk1"

    init{
        partedParser = PartedParserRunnable()
    }

    override fun getContext():Context {
        return context
    }

    override fun onCreateView(inflater:LayoutInflater, container:ViewGroup?, savedInstanceState:Bundle?):View? {
        if (rootView == null) {
            rootView = inflater.inflate(R.layout.partition_scheme, container, false) as RelativeLayout
            onViewFirstCreated()
        }
        rootView!!.startAnimation(MainActivity.mFadeIn)
        return rootView
    }

    @SuppressLint("InflateParams")
    private fun onViewFirstCreated() {
        partitionSwipeRefresh = rootView!!.findViewById(R.id.partition_swipe_refresh) as SwipeRefreshLayout
        partitionSwipeRefresh!!.setOnRefreshListener {
            partitionSwipeRefresh!!.isRefreshing = true
            refresh(1)
        }
        partitionListView = rootView!!.findViewById(R.id.partition_list_view) as ListView
        if (blockDevicesList == null) {
            partitionSwipeRefresh!!.isRefreshing = true
            blockDevicesList = ArrayList()
            newAdapter()
            refresh(0)
        } else newAdapter()

        mainBlockDevices = ArrayList()
        if (File(blockDev0).exists()) mainBlockDevices!!.add(blockDev0)
        if (File(blockDev1).exists()) mainBlockDevices!!.add(blockDev1)
            partitionListView!!.adapter = adapter
        headerView = (getContext() as AppCompatActivity).layoutInflater.inflate(R.layout.partition_view_header, null)
        spinnerAdapter = ArrayAdapter(getContext(), android.R.layout.simple_list_item_1, mainBlockDevices!!)
        (headerView!!.findViewById(R.id.part_disk_spinner) as Spinner).adapter = spinnerAdapter
        headerView!!.visibility = View.GONE
        partitionListView!!.addHeaderView(headerView)
    }

    private fun newAdapter() {
        adapter = PartitionListAdapter(getContext(), R.layout.partition_list_item, blockDevicesList!!)
    }

    private fun refresh(type:Int) {
        blockDevicesList!!.clear()
        adapter!!.notifyDataSetChanged()
        if (!(getContext() as MainActivity).backgroundThreadisRunning)
            runInBackground(partedParser.setType(type))
        else
            Thread(partedParser).start()
    }

    private fun onLoadingCompleted(partedData:String) {
        blockDevicesList!!.clear()
        blockDevicesList!!.addAll(parseParted(partedData))
        runOnUiThread(Runnable {
            headerView!!.visibility = View.VISIBLE
            adapter!!.notifyDataSetChanged()
            partitionSwipeRefresh!!.isRefreshing = false
            setHasOptionsMenu(true)
            mainBlockDevices!!.removeAt(0)
            mainBlockDevices!!.add(0, blockDev0 + " : " + Utils.getConventionalSize(diskSize))
            spinnerAdapter!!.notifyDataSetChanged()
            (headerView!!.findViewById(R.id.part_table) as TextView).text = partitionTable
            (headerView!!.findViewById(R.id.disk_model) as TextView).text = storageModel
            (headerView!!.findViewById(R.id.sector_size) as TextView).text = selectSize
        })
    }

    private fun runInBackground(action:Runnable) {
        (getContext() as MainActivity).runInBackground(action)
    }

    private fun runOnUiThread(action:Runnable) {
        (getContext() as MainActivity).runOnUiThread(action)
    }

    private fun writeToFile(data:String, fileName:String, context:Context) {
        val outputStream:FileOutputStream
        try {
            outputStream = context.openFileOutput(fileName, Context.MODE_PRIVATE)
            outputStream.write(data.toByteArray())
            outputStream.close()
        }
        catch (ex:Exception) {
            Log.e(MainActivity.TAG, "File write failed: $ex")
        }
    }

    override fun onCreateOptionsMenu(menu:Menu, inflater:MenuInflater) {
        inflater.inflate(R.menu.menu_part_unit, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item:MenuItem):Boolean {
        for (data in blockDevicesList!!)
            data.setSizeUnit(item.itemId)
        adapter!!.notifyDataSetChanged()
        return super.onOptionsItemSelected(item)
    }

    private fun parseParted(str:String):ArrayList<BlockDeviceListData> {
        val devicesList = ArrayList<BlockDeviceListData>()
        for (tmp in str.split(("\n").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
            if (storageModel == null && tmp.contains("Model"))
            storageModel = tmp.replace("Model: ", "").replace("Model", "")
            else if (tmp.contains("Disk ")) {
                try {
                    disk = File(tmp.split((":").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0].split((" ").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1])
                    diskSize = java.lang.Long.parseLong(tmp.split((":").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].replace("B", "").replace(" ", ""))
                } catch (ex:NullPointerException) {
                    Log.e(MainActivity.TAG, ex.toString())
                }
            }
            else if (tmp.contains("Sector size")) {
                try {
                    selectSize = tmp.split((":").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].replace(" ", "")
                } catch (ex:NullPointerException) {
                    Log.e(MainActivity.TAG, ex.toString())
                }
            } else if (tmp.contains("Partition Table")) {
                try {
                    partitionTable = tmp.split((":").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].replace(" ", "")
                } catch (ex:NullPointerException) {
                    Log.e(MainActivity.TAG, ex.toString())
                }
            }
            else if (tmp.contains("Number "))
                for (test in str.split((tmp).toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1].split(("\n").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
                    if (test.length > 10) {
                        var i = 0
                        val data = BlockDeviceListData()
                        for (another in test.split((" ").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
                        if (another != "") {
                            when (i) {
                                0 -> data.setBlock(File((disk).toString() + "p" + Integer.parseInt(another)))
                                1 -> data.setStart(java.lang.Long.parseLong(another.split(("B").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]))
                                2 -> data.setEnd(java.lang.Long.parseLong(another.split(("B").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]))
                                3 -> data.setSize(java.lang.Long.parseLong(another.split(("B").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]))
                                else -> {
                                    try {
                                        if (data.getName() != null) data.setType(data.getName()!!)
                                        else data.setType("")
                                    } catch (ex:Exception) {
                                        data.setType("")
                                    }
                                    data.setName(another)
                                }
                            }
                            i++
                        }
                        if (data.getBlock() != null) devicesList.add(data)
                    }
        }
        return devicesList
    }

    private inner class PartedParserRunnable:Runnable {
        private var type = 0
        fun setType(type:Int):Runnable {
            this.type = type
            return this
        }

        override fun run() {
            val dataFile = File((getContext().filesDir).toString() + "/partition_scheme.info")
            var scheme = ""
            if (dataFile.exists() && type == 0) {
                try {
                    val reader = BufferedReader(InputStreamReader(FileInputStream(dataFile)))
                    var tmp = reader.readLine()
                    while (tmp!= null) {
                        scheme += tmp + "\n"
                        tmp = reader.readLine()
                    }
                    onLoadingCompleted(scheme)
                } catch (ex:IOException) {
                    Log.e("Exception", "Failed to read " + dataFile.absolutePath, ex)
                }
            } else {
                if (dataFile.exists()) dataFile.delete()
                MainActivity.rootSession!!.addCommand((getContext().filesDir).toString() + "/common/partition_scheme.sh " + MainActivity.TOOL + " 'b'", 2323) { _, _, output ->
                    writeToFile(Utils.getString(output), "partition_scheme.info", context)
                    onLoadingCompleted(Utils.getString(output))
                }
            }
        }
    }

    companion object {

        private var blockDevicesList:ArrayList<BlockDeviceListData>? = null
    }
}

